docs/src/design/proposals/APIRedesign.md

```diff
@@ -0,0 +1,1050 @@
+developer console.
+
+This is acceptable behavior for truly unrecoverable missing
+dependencies, but in many cases it would be preferable to allow a
+given type of composite service to define some failure behavior
+when no service of an appropriate type is available.
+
+To address this:
+
+* Provide an interface (preferably
+  [imperative](#bundle-Declarations-in-javascript))
+  for declaring composite services, independent of any implementation
+  of an aggregator/decorator/provider. This allows the framework
+  layer to distinguish between unimplemented dependencies (which
+  could have defined failover strategies) from undefined dependencies
+  (which cannot.)
+* Provide a default strategy for service composition that picks
+  the highest-priority provider, and logs an error (and fails to
+  satisfy) if no providers have been defined.
+* Allow this aggregation strategy to be overridden, much as one
+  can declare aggregators currently. However, these aggregators should
+  get empty arrays when no providers have been registered (instead of
+  being ignored), at which point they can decide how to handle this
+  situation (graceful failure when it's possible, noisy errors when
+  it is not.)
+
+### Benefits
+
+* Allows for improved robustness and fault tolerance.
+* Makes service declarations explicit, reducing "magic."
+
+### Detriments
+
+* Requires the inclusion of software units which define services,
+  instead of inferring their existence (slight increase in amount
+  of code that needs to be written.)
+* May result in harder-to-understand errors when overridden
+  composition strategies do not failover well (that is, when they
+  do need at least implementation, but fail to check for this.)
+
+## Plugins as Angular Modules
+
+Do away with the notion of bundles entirely; use Angular modules
+instead. Registering extensions or components of composite services
+would then be handled by configuring a provider; reusable classes
+could be exposed by the platform for these.
+
+Example (details are flexible, included for illustrative purposes):
+
+```javascript
+var mctEdit = angular.module('mct-edit', ['ng', 'mct']);
+
+// Expose a new extension category
+mctEdit.provider('actionRegistry', ExtensionCategoryProvider);
+
+// Expose a new extension
+mctEdit.config(['actionRegistryProvider', function (arp) {
+    arp.register(EditPropertiesAction);
+}])
+
+return mctEdit;
+```
+
+Incompatible with proposal to
+(expose no third-party APIs)[#expose-no-third-party-apis]; Angular
+API would be ubiquitously exposed.
+
+### Benefits
+
+* Removes a whole category of API (bundle definitions), reducing
+  learning curve associated with the software.
+* Closer to Angular style, reducing disconnect between learning
+  Angular and learning Open MCT Web (reducing burden of having
+  to learn multiple paradigms.)
+* Clarifies "what can be found where" (albeit not perfectly)
+  since you can look to module dependencies and follow back from there.
+
```

Comment:
Not necessarily a benefit or detriment, but as of Angular 1.5 there are some interesting additions to angular.Module that are relevant here:

* [`angular.Module.decorator`](https://docs.angularjs.org/api/ng/type/angular.Module#decorator) is a thing now. This looks like an alias of the existing [`$provide.decorator`](https://docs.angularjs.org/api/auto/service/$provide#decorator), which was easy to miss. Doesn't have the priority-ordering effect of our composite services, but not sure we need that (although ordering does matter for some decorators - policy, for instance, can get subverted if it's enforced at the wrong point in a chain.)
* [`angular.Module.component`](https://docs.angularjs.org/api/ng/type/angular.Module#component) has also been added as a short-hand for registering Web Component-like directives. This means that what we currently call `components` will either need to be renamed, or go away.

Only things that stand out as missing relative to our current framework layer are (a) priority ordering and (b) many-to-one dependencies.